internal/provider/resource_cloudflare_web3hostname.go

```diff
@@ -0,0 +1,92 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceCloudflareWeb3Hostname() *schema.Resource {
+	return &schema.Resource{
+		Schema:        resourceCloudflareWeb3HostnameSchema(),
+		CreateContext: resourceCloudflareWeb3HostnameCreate,
+		ReadContext:   resourceCloudflareWeb3HostnameRead,
+		UpdateContext: resourceCloudflareWeb3HostnameUpdate,
+		DeleteContext: resourceCloudflareWeb3HostnameDelete,
+	}
+}
+
+func resourceCloudflareWeb3HostnameCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+
+	hostname, err := client.CreateWeb3Hostname(ctx, cloudflare.Web3HostnameCreateParameters{
+		ZoneID:      d.Get("zone_id").(string),
+		Name:        d.Get("name").(string),
+		Target:      d.Get("target").(string),
+		Description: d.Get("description").(string),
+		DNSLink:     d.Get("dnslink").(string),
+	})
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error creating web3hostname %q: %w", d.Get("name").(string), err))
+	}
+
+	d.SetId(hostname.ID)
+
+	return resourceCloudflareWeb3HostnameRead(ctx, d, meta)
+}
+
+func resourceCloudflareWeb3HostnameRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+
+	hostname, err := client.GetWeb3Hostname(ctx, cloudflare.Web3HostnameDetailsParameters{
+		ZoneID:     d.Get("zone_id").(string),
+		Identifier: d.Id(),
+	})
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading web3hostname %q: %w", d.Id(), err))
+	}
+
+	d.SetId(hostname.ID)
+
+	return resourceCloudflareWeb3HostnameRead(ctx, d, meta)
```

Comment:
you'll want to return `nil` here otherwise you'll end up recursively calling itself.
```suggestion
	return nil
```